### PR TITLE
feat(api): Add initial skeleton for deprecation decorator

### PR DESCRIPTION
**Pull request by AniketDas-Tekky**
The decorator will eventually handle all the overhead for a deprecated endpoint including
* Metric tracking
* Brownout
* Deprecation Headers

Later PRs will introduce crontab logic for brownout periods and unit tests

### CODE REVIEW

**Review by wedamija — COMMENTED** [review]
How will these deprecation dates work for ST? Will they be on the same schedule?

I assume that we won't perform brownouts for self-hosted and that these endpoints will just disappear in the version after the deprecation date.

I'd like to see tests for this that validate that the variables on the `request` are set as expected, and that endpoints are blocked appropriately.

**Comment by AniketDas-Tekky** [comment]
> How will these deprecation dates work for ST? Will they be on the same schedule?
> 
> I assume that we won't perform brownouts for self-hosted and that these endpoints will just disappear in the version after the deprecation date.
> 
> I'd like to see tests for this that validate that the variables on the `request` are set as expected, and that endpoints are blocked appropriately.

Good catch, yeah I'll disable deprecation for self hosted.

File: src/sentry/api/helpers/deprecation.py [hunk 1]
```diff
@@ -46,14 +55,18 @@ def decorator(func):
         @functools.wraps(func)
         def endpoint_method(self, request: Request, *args, **kwargs):
 
+            # Don't do anything for deprecated endpoitns on self hosted
```
Comment (markstory) [inline src/sentry/api/helpers/deprecation.py:66]:
```suggestion
            # Don't do anything for deprecated endpoints on self hosted
```

File: src/sentry/api/helpers/deprecation.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+from rest_framework.status import HTTP_410_GONE as GONE
+
+from sentry.utils.settings import is_self_hosted
+
+BROWNOUT_LENGTH = timedelta(days=30)
+GONE_MESSAGE = {"message": "This API no longer exists."}
+DEPRECATION_HEADER = "X-Sentry-Deprecation-Date"
+SUGGESTED_API_HEADER = "X-Sentry-Replacement-Endpoint"
+
+# TODO: Make these configurable from redis
```
Comment (markstory) [inline src/sentry/api/helpers/deprecation.py:18]: You could use `sentry.options` for this. Ops has tooling to update options already.

Comment (AniketDas-Tekky) [reply src/sentry/api/helpers/deprecation.py:18]: Yeah, I was going to do more of the brownout logic in a followup PR. I need to think about what sort of format the brownout schedule should be.

File: src/sentry/api/helpers/deprecation.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+            # Don't do anything for deprecated endpoints on self hosted
+            if is_self_hosted():
+                return func(self, request, *args, **kwargs)
+
+            now = datetime.now(timezone.utc)
+            # TODO: Need a better way to flag a request on a deprecated endpoint
+            # _track_deprecated_metrics(request, deprecation_date)
+
+            if now > deprecation_date and _should_be_blocked(deprecation_date, now):
```
Comment (wmak) [inline src/sentry/api/helpers/deprecation.py:79]: (For another PR perhaps) Could we add something here so the endpoint could implement custom logic for whether an endpoint should be blocked? I'd like to be able to add a killswitch via Flagr for example

**Review by wmak — APPROVED** [review]
LGTM

**Pull request merged**